### PR TITLE
librgw: support symbolic link

### PR DESCRIPTION
**Pull request by taoCH**
Signed-off-by: Tao Chen <chentao@umcloud.com>

Librgw currently supports file, directory as an object, I think we could also make a symbolic link as an object too.

### CODE REVIEW

File: src/include/rados/rgw_file.h [hunk 1]
```diff
@@ -178,6 +179,15 @@ int rgw_create(struct rgw_fs *rgw_fs, struct rgw_file_handle *parent_fh,
 	       struct rgw_file_handle **fh, uint32_t posix_flags,
 	       uint32_t flags);
 
+/*
+  create a symbolic link
+ */
+#define RGW_CREATELINK_FLAG_NONE     0x0000
```
Comment (ZVampirEM77) [inline src/include/rados/rgw_file.h]: This macro is unused? Clean up it please~

**Comment by taoCH** [comment]
@mattbenjamin  Hi Matt, could you please review this commit?

File: src/rgw/rgw_file.cc [hunk 2]
```diff
@@ -1668,6 +1765,35 @@ int rgw_create(struct rgw_fs *rgw_fs, struct rgw_file_handle *parent_fh,
   return get<1>(fhr);
 } /* rgw_create */
 
+/*
+  create a symbolic link
+ */
+int rgw_createlink(struct rgw_fs *rgw_fs, struct rgw_file_handle *parent_fh,
```
Comment (mattbenjamin) [inline src/rgw/rgw_file.cc]: s/createlink/symlink/;

File: src/rgw/rgw_file.h [hunk 3]
```diff
@@ -316,7 +317,10 @@ namespace rgw {
 	if (flags & FLAG_DIRECTORY) {
 	  fh.fh_type = RGW_FS_TYPE_DIRECTORY;
 	  variant_type = directory();
-	} else {
+        } else if(flags & FLAG_SYMBOLIC_LINK) {
+	  fh.fh_type = RGW_FS_TYPE_SYMBOLIC_LINK;
```
Comment (mattbenjamin) [inline src/rgw/rgw_file.h]: ok

File: src/rgw/rgw_file.h [hunk 4]
```diff
@@ -336,6 +340,10 @@ namespace rgw {
 	break;
       case RGW_FS_TYPE_FILE:
 	state.unix_mode = RGW_RWMODE|S_IFREG;
+        break;
+      case RGW_FS_TYPE_SYMBOLIC_LINK:
+        state.unix_mode = RGW_RWMODE|S_IFLNK;
```
Comment (mattbenjamin) [inline src/rgw/rgw_file.h]: ok

File: src/rgw/rgw_file.h [hunk 5]
```diff
@@ -386,6 +394,10 @@ namespace rgw {
 	  break;
 	case RGW_FS_TYPE_FILE:
 	  state.unix_mode = st->st_mode|S_IFREG;
+          break;
+        case RGW_FS_TYPE_SYMBOLIC_LINK:
+          state.unix_mode = st->st_mode|S_IFLNK;
```
Comment (mattbenjamin) [inline src/rgw/rgw_file.h]: ok

File: src/rgw/rgw_file.h [hunk 6]
```diff
@@ -429,6 +441,13 @@ namespace rgw {
 	st->st_blksize = 4096;
 	st->st_size = state.size;
 	st->st_blocks = (state.size) / 512;
+        break;
+      case RGW_FS_TYPE_SYMBOLIC_LINK:
+	st->st_nlink = 1;
```
Comment (mattbenjamin) [inline src/rgw/rgw_file.h]: yeah, ok;

File: src/rgw/rgw_file.h [hunk 7]
```diff
@@ -540,6 +559,7 @@ namespace rgw {
     bool is_object() const { return !is_bucket(); }
     bool is_file() const { return (fh.fh_type == RGW_FS_TYPE_FILE); }
     bool is_dir() const { return (fh.fh_type == RGW_FS_TYPE_DIRECTORY); }
+    bool is_link() const { return (fh.fh_type == RGW_FS_TYPE_SYMBOLIC_LINK); }
```
Comment (mattbenjamin) [inline src/rgw/rgw_file.h]: ack

File: src/rgw/rgw_file.h [hunk 8]
```diff
@@ -637,6 +657,9 @@ namespace rgw {
       DECODE_START(2, bl);
       uint32_t fh_type;
       ::decode(fh_type, bl);
+      if (fh.fh_type != fh_type &&
```
Comment (mattbenjamin) [inline src/rgw/rgw_file.h]: @taoCH for clarity, could you put the subexpressions in parentheses, i.e
if ((fh.fh_type != fh_type) &&
and ff.?

File: src/rgw/rgw_file.h [hunk 9]
```diff
@@ -1129,6 +1152,9 @@ namespace rgw {
 
     MkObjResult create(RGWFileHandle* parent, const char *name, struct stat *st,
 		      uint32_t mask, uint32_t flags);
+    
+    MkObjResult createlink(RGWFileHandle* parent, const char *name,
```
Comment (mattbenjamin) [inline src/rgw/rgw_file.h]: s/createlink/symlink/;

File: src/include/rados/rgw_file.h [hunk 10]
```diff
@@ -39,6 +39,7 @@ enum rgw_fh_type {
   RGW_FS_TYPE_NIL = 0,
   RGW_FS_TYPE_FILE,
   RGW_FS_TYPE_DIRECTORY,
+  RGW_FS_TYPE_SYMBOLIC_LINK,
 };
 
 /*
```
Comment (mattbenjamin) [inline src/include/rados/rgw_file.h]: @taoCH #define LIBRGW_FILE_VER_EXTRA needs to increment as you're extending the interface

**Comment by taoCH** [comment]
@mattbenjamin  Thanks for reviewing and your advice, I will modify this PR following you advice when #19358 been merged

**Comment by mattbenjamin** [comment]
@taoCH ready for rebase

**Comment by taoCH** [comment]
@mattbenjamin  code refreshed

**Comment by taoCH** [comment]
@mattbenjamin  Hi, I have modified the code as your advice, could pls review this?

**Comment by taoCH** [comment]
@mattbenjamin Hi, I have modified the code as your advice, could pls review this?

**Comment by taoCH** [comment]
ok, code rebased

**Comment by yuriw** [comment]
[wip-yuri4-testing-2018-11-15-2120](https://trello.com/c/rED6z2kY/473-wip-yuri4-testing-2018-11-15-2120)

**Pull request merged**